pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,295 @@
+package stats
+
+import (
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+)
+
+const (
+	bucketDuration    = uint64(1e10) // 10s
+	oldestBucketStart = 20 * time.Second
+)
+
+const (
+	// set on a stat payload containing only distributions post aggregation
+	keyDistributions = "distributions"
+	// set on a stat payload containing counts (hit/error/duration) post aggregation
+	keyCounts = "counts"
+)
+
+// ClientStatsAggregator aggregates client stats payloads on 1s buckets.
+// If a single payload is received on a bucket, this Aggregator is a passthrough.
+// If two or more payloads collide, their counts will be aggregated into one bucket.
+// Multiple payloads will be sent:
+// - Original payloads with their distributions will be sent with counts zeroed.
+// - A single payload with the bucket aggregated counts will be sent.
+// This ensures that all counts will have at most one point per second per agent for a specific granularity.
+// While distributions are not tied to the agent.
+type ClientStatsAggregator struct {
+	In      chan pb.ClientStatsPayload
+	out     chan pb.StatsPayload
+	buckets map[uint64]*bucket // buckets used to aggregate client stats
+
+	flushTicker   *time.Ticker
+	oldestTs      uint64
+	agentEnv      string
+	agentHostname string
+
+	exit chan struct{}
+	done chan struct{}
+}
+
+// NewClientStatsAggregator initializes a new aggregator ready to be started
+func NewClientStatsAggregator(conf *config.AgentConfig, out chan pb.StatsPayload) *ClientStatsAggregator {
+	return &ClientStatsAggregator{
+		flushTicker:   time.NewTicker(time.Second),
+		In:            make(chan pb.ClientStatsPayload, 10),
+		buckets:       make(map[uint64]*bucket, 20),
+		out:           out,
+		agentEnv:      conf.DefaultEnv,
+		agentHostname: conf.Hostname,
+		oldestTs:      uint64(time.Now().Add(-oldestBucketStart).Unix()),
+		exit:          make(chan struct{}),
+		done:          make(chan struct{}),
+	}
+}
+
+// Start starts the aggregator.
+func (a *ClientStatsAggregator) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
+		for {
+			select {
+			case t := <-a.flushTicker.C:
+				a.flushOnTime(t)
+			case input := <-a.In:
+				a.add(time.Now(), input)
```

Comment:
Testing might be easier if this struct has a field like
```
clock func() time.Time
```
That you call instead of explicitly calling `time.Now()`. The value for this field can be `time.Now` by default, but you can override in tests.

Reply:
I couldn't find how to simplify my tests with the clock and did not change this

Reply:
It can be helpful if you want to validate behavior under certain timing conditions.